src/main/scala/mesosphere/marathon/core/storage/impl/BasePersistenceStore.scala

```diff
@@ -0,0 +1,141 @@
+    val storageId = ir.toStorageId(k, Some(version))
+    lockManager.executeSequentially(k.toString) {
+      rawDelete(ir.toStorageId(k, Some(version)), version)
+    }
+  }
+
+  protected def rawDeleteAll(k: K): Future[Done]
+
+  final override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    val storageId = ir.toStorageId(k, None)
+    lockManager.executeSequentially(k.toString) {
+      rawDeleteAll(ir.toStorageId(k, None))
+    }
+  }
+
+  protected[storage] def rawGet(k: K): Future[Option[Serialized]]
+
+  override def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = async {
+    val storageId = ir.toStorageId(id, None)
+    await(rawGet(storageId)) match {
+      case Some(v) =>
+        Some(await(Unmarshal(v).to[V]))
+      case None =>
+        None
+    }
+  }
+
+  override def get[Id, V](id: Id, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = async {
+    val storageId = ir.toStorageId(id, Some(version))
+    await(rawGet(storageId)) match {
+      case Some(v) =>
+        Some(await(Unmarshal(v).to[V]))
+      case None =>
+        None
+    }
+  }
+
+  protected def deleteOld(k: K, maxVersions: Int): Future[Done] = async {
+    val versions = await(rawVersions(k).toMat(Sink.seq)(Keep.right).run()).sortBy(_.toEpochSecond)
+    // we always store the current version (twice), once as a versioned node and once as the current one.
+    val numToDelete = versions.size - maxVersions - 1
+    if (numToDelete > 0) {
+      val deletes = versions.take(numToDelete).map(v => rawDelete(k, v))
+      await(Future.sequence(deletes))
+    }
+    Done
+  }
+
+  protected def rawStore[V](k: K, v: Serialized): Future[Done]
+
+  override def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done] = {
+    val storageId = ir.toStorageId(id, None)
+    lockManager.executeSequentially(id.toString) {
+      async {
+        val serialized = await(Marshal(v).to[Serialized])
+        val (_, _) = (
+          await(rawStore(storageId, serialized)),
+          await(rawStore(ir.toStorageId(id, Some(ir.version(v))), serialized)))
+        await(deleteOld(storageId, ir.maxVersions))
+        Done
+      }
+    }
+  }
+
+  override def store[Id, V](id: Id, v: V,
+    version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized]): Future[Done] = {
+
+    val storageId = ir.toStorageId(id, Some(version))
+    val currentId = ir.toStorageId(id, None)
+    lockManager.executeSequentially(id.toString) {
+      async {
+        val serialized = await(Marshal(v).to[Serialized])
+        await(rawGet(currentId)) match {
```

Comment:
This will store the value as current version, if not existent.
Could be surprising - If I say store this entry in this version, I would not expect to store it also without version.